core/server/worker/src/main/java/alluxio/worker/block/BlockMasterClient.java

```diff
@@ -269,6 +274,44 @@ public void acquireRegisterLeaseWithBackoff(
         workerId, retry.getAttemptCount(), response);
   }
 
+  /**
+   * Registers with the Alluxio master. This should be called before the
+   * {@link BlockMasterSync#registerWithMaster}, The method used to check whether the worker
+   * can register with the current cluster.
+   * If it passes check ClusterId and workerId be returned
+   * @param clusterId the cluster id of the worker
+   * @param address the worker WorkerNetAddress
+   * @param blocksNum The number of blocks in the worker
+   * @return an optional command for the worker to execute
+   */
+  public GetWorkerIdPResponse preRegisterWithMaster(String clusterId,
+      final WorkerNetAddress address, int blocksNum)
+      throws AlluxioStatusException {
+
+    // For compatibility, reuse the RPC of getWorkerID.
+    GetWorkerIdPRequest request = GetWorkerIdPRequest.newBuilder()
+        .setClusterId(clusterId).setWorkerNetAddress(GrpcUtils.toProto(address))
+        .setBlocksNum(blocksNum).build();
+
+    return retryRPC(() -> {
+      // For compatibility, reuse the RPC of getWorkerID.
+      GetWorkerIdPResponse response = mClient.getWorkerId(request);
+      // For compatibility,  If preRegister information is included,
+      // the Master will set hasExtendedRegisterInfo
+      if (response.hasClusterId()) {
```

Comment:
i think you are doing this compatibility check anyway, so you can just merge this perRegisterWithMaster() with the existing getWorkerId()?